### PR TITLE
Record inner instruction stack height

### PR DESCRIPTION
**Pull request by jstarry**
#### Problem
The inner instruction stack height isn't recorded after transaction processing so clients can't determine if an inner instruction was invoked by another inner instruction.

#### Summary of Changes
Builds on https://github.com/solana-labs/solana/pull/28427

- Record `stack_height` in a new `InnerInstruction` protobuf message type for each inner instruction
- Add a new `stack_height` field to instruction data types returned by the RPC API

Related to https://github.com/solana-labs/solana/issues/27736

### CODE REVIEW

**Comment by Lichtso** [comment]
We can't change `CompiledInstruction` and thus have to add a wrapper `InnerInstruction`, or what is it for?

**Comment by jstarry** [comment]
> We can't change CompiledInstruction and thus have to add a wrapper InnerInstruction, or what is it for?

Yes, exactly. Can't change `CompiledInstruction` in the sdk since it's used for transaction serialization. `InnerInstruction` is just a wrapper type.. it doesn't have to be specific to inner instructions but it's not necessary to record the stack height for transaction level instructions.

File: storage-proto/proto/confirmed_block.proto [hunk 1]
```diff
@@ -72,7 +72,18 @@ message TransactionError {
...
+
+message InnerInstruction {
+    uint32 program_id_index = 1;
+    bytes accounts = 2;
+    bytes data = 3;
+
+    // Invocation stack height of an inner instruction.
+    // Available since Solana v1.14.6
+    // Set to `None` for txs executed on earlier versions.
+    optional uint32 stack_height = 4;
```
Comment (Lichtso) [inline storage-proto/proto/confirmed_block.proto:86]: `uint32` is that really necessary, the runtime itself only used 8 bits for the stack height.

Comment (jstarry) [reply storage-proto/proto/confirmed_block.proto:86]: Smallest size in protobuf is u32

File: runtime/src/bank.rs [hunk 2]
```diff
@@ -489,7 +489,16 @@ pub type TransactionBalances = Vec<Vec<u64>>;
...
-pub type InnerInstructions = Vec<CompiledInstruction>;
+pub type InnerInstructions = Vec<InnerInstruction>;
+
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub struct InnerInstruction {
+    pub instruction: CompiledInstruction,
+    /// Invocation stack height of this instruction. Instruction stack height
+    /// starts at 1 for transaction instructions. This field is set to `None`
+    /// for transactions that were processed prior to stack height tracking.
+    pub stack_height: Option<u32>,
```
[the remark this answers is not in the document]
Comment (jstarry) [reply runtime/src/bank.rs]: Done

**Review by Lichtso — APPROVED** [review]

**Pull request merged**